Dockerfile

```diff
@@ -65,26 +37,45 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
     for f in $(find bin -executable -type f); do xx-verify $f; done
 
 FROM --platform=${BUILDPLATFORM} scratch AS spire-base
-WORKDIR /opt/spire
+# For users that wish to run SPIRE containers as a non-root user,
+# provide a default unprivileged user such that the default paths
+# that SPIRE will try to read from, write to, and create at runtime
+# can be given the correct file ownership/permissions at build time.
+ARG spireuid=1000
+ARG spiregid=1000
 CMD []
-COPY --link --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
+COPY --link --from=builder --chown=${spireuid}:${spiregid} --chmod=777 /empty-dir /opt/spire
```

Comment:
Could we use separate stages for creating the directories and then apply the permissions when they are copied to the final images? Something like this:

```
# Setting up SPIRE Server directories
FROM alpine as prep-server
RUN mkdir -p /opt/spire/bin \
    /etc/spire/server \
    /run/spire/server/private \
    /tmp/spire-server/private \
    /var/lib/spire/server

# Setting up SPIRE Agent directories
FROM alpine as prep-agent
RUN mkdir -p /opt/spire/bin \
    /etc/spire/agent \
    /run/spire/agent/public \
    /tmp/spire-agent/public \
    /var/lib/spire/agent

# SPIRE Server
FROM spire-base AS spire-server
ARG spireuid=1000
ARG spiregid=1000
USER ${spireuid}:${spiregid}
COPY --link --from=prep-server --chown=${spireuid}:${spiregid} --chmod=755 / /
COPY --link --from=builder --chown=${spireuid}:${spiregid} --chmod=755 /spire/bin/static/spire-server /opt/spire/bin/
ENTRYPOINT ["/opt/spire/bin/spire-server", "run"]

# Similar configuration for SPIRE Agent
```

Reply:
Also `spire-base` could be just:

```
FROM --platform=${BUILDPLATFORM} scratch AS spire-base
COPY --link --from=builder --chown=root:root --chmod=755 /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
WORKDIR /opt/spire
```

Reply:
Thank you @maxlambrecht for the suggestions, I think that they make sense.
I've updated the Dockerfile accordingly.